pkg/build/api/validation/validation.go

```diff
@@ -151,50 +161,27 @@ func hasProxy(source *buildapi.GitBuildSource) bool {
 
 func validateSource(input *buildapi.BuildSource) fielderrors.ValidationErrorList {
 	allErrs := fielderrors.ValidationErrorList{}
-	switch input.Type {
-	case buildapi.BuildSourceGit:
-		if input.Git == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("git"))
-		} else {
-			allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
-		}
+
+	if input.Git != nil {
+		allErrs = append(allErrs, validateGitSource(input.Git).Prefix("git")...)
 		if input.Binary != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when type is Git"))
-		}
-	case buildapi.BuildSourceBinary:
-		if input.Binary == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("binary"))
-		} else {
-			allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
-		}
-		if input.Dockerfile != nil {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("binary", "", "may not be set when git also set"))
 		}
+	}
+	if input.Binary != nil {
+		allErrs = append(allErrs, validateBinarySource(input.Binary).Prefix("binary")...)
 		if input.Git != nil {
-			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when type is Binary"))
-		}
-	case buildapi.BuildSourceDockerfile:
-		if input.Dockerfile == nil {
-			allErrs = append(allErrs, fielderrors.NewFieldRequired("dockerfile"))
-		} else {
-			allErrs = append(allErrs, validateDockerfile(*input.Dockerfile)...)
+			allErrs = append(allErrs, fielderrors.NewFieldInvalid("git", "", "may not be set when binary is also set"))
```

Comment:
If `Git` and `Binary` are both given, `validateSource` will return:

```go
fielderrors.ValidationErrorList{
	fielderrors.NewFieldInvalid("binary", "", "may not be set when git also set"),
	fielderrors.NewFieldInvalid("git", "", "may not be set when binary is also set"),
	// and perhaps even more...
	fielderrors.NewFieldInvalid("uri", git.URI, "uri is not a valid url"),
}
```

IMO, it should return:

```go
fielderrors.ValidationErrorList{
	fielderrors.NewFieldInvalid("", "", git and binary sources may not be used together),
}
```